chore: Docker Compose import milestone 1 - library setup and basic service conversion

rclinard-amzn opened the pull request:
<!-- Provide summary of changes -->

Note: This is targeting the `experiment/docker-compose-import` branch. It is not targeting the mainline release branch. 
Also note that most of the added code is unit tests - there's <250 lines of actual functional code.

This PR implements the first milestone of the Docker Compose import experiment. We set up the library and handle the Compose concepts that mostly mirror 1:1 with Copilot fields.

In the next PR, I will add detection for unsupported Compose fields that this experiment won't support.

Relates to https://github.com/aws/copilot-cli/issues/1612

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

rclinard-amzn committed:
chore: import compose-go library

rclinard-amzn committed:
chore: implement conversion to Copilot's image manifest field

rclinard-amzn committed:
chore: Add tests for svc_image.go

rclinard-amzn committed:
chore: improved nil handling in svc_image.go

rclinard-amzn committed:
chore: implement backend & lbws conversion, begin unit tests for that too

rclinard-amzn committed:
chore: warn on build.labels being set

rclinard-amzn committed:
chore: fix handling of Extensions field in svc_image.go

rclinard-amzn committed:
chore: fix handling of Extensions field

rclinard-amzn committed:
chore: Add comprehensive test case for backend service conversion

rclinard-amzn committed:
chore: remove LBWS conversion until we figure out how to avoid massive code duplication

rclinard-amzn committed:
chore: flesh out svc_test.go

rclinard-amzn committed:
chore: do not export fields in unsupported.go, comment as unused

rclinard-amzn committed:
chore: test for detection of extension fields

rclinard-amzn commented:
One issue that needs to be addressed after this PR - how do we handle LBWS conversion without duplicating tons of code? The LBWS manifest is almost exactly the same as the Backend Service manifest, but they use two different data structures in our codebase because the Port field is optional on Backend services. Alternatively, we can just not add the same level of unit tests for the LBWS conversion function since the logic should otherwise be the same.

rclinard-amzn committed:
chore: add license headers

rclinard-amzn committed:
chore: comment out unused unsupported field lists until the next PR

fixes staticcheck failing for now

rclinard-amzn committed:
chore: remove ineffectual assignments in TestConvertImageConfigNil

huanjani commented on internal/pkg/docker/dockercompose/svc.go:
```diff
@@ -0,0 +1,104 @@
+	"time"
+)
+
+func convertBackendService(service *types.ServiceConfig, port uint16) (*manifest.BackendServiceConfig, IgnoredKeys, error) {
```
Just curious how the returned `IgnoredKeys` will be handled.

rclinard-amzn replied:
I intend for it to be passed through up to the point where the user is interacting with the software, so that an appropriate warning message can be displayed to them, without disrupting conversion. The fatal keys, however, generate an error instead.

huanjani replied:
👍🏼 
Just a note: let's build with conversion extensability in mind; as Docker Compose or we add/change/(deprecate) fields, it should be easy enough to adapt in the conversion code.

huanjani commented on internal/pkg/docker/dockercompose/svc_image.go:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	// note: Compose allows both build & image to be specified, but image takes precedence.
+	//       in Copilot, however, build & image are mutually exclusive.
```
nit:
```suggestion
	// In Copilot, however, build and image are mutually exclusive.
```

CaptainCarpensir commented:
> One issue that needs to be addressed after this PR - how do we handle LBWS conversion without duplicating tons of code? The LBWS manifest is almost exactly the same as the Backend Service manifest, but they use two different data structures in our codebase because the Port field is optional on Backend services. Alternatively, we can just not add the same level of unit tests for the LBWS conversion function since the logic should otherwise be the same.

Unit tests are not good to skip even in these cases because they're partly needed to alert people when things break. Because they may be similar code it may be attractive to skip writing them, but they can save you lots of time debugging, and help you find bugs before they're sent to production. It's not just to test the logic of code, but also to make sure that the code is what you think it is, typos can break a surprising amount of stuff 😄

rclinard-amzn commented:
That's a good point! I think that once the Backend Service handling is fleshed out in a PR or two, the LBWS handling can be copy pasted from that, instead of having to copy paste every time I make a logic change to Backend Service handling. That should amortize the cost of copy-pasting and avoid introducing minor deviations along the way!

huanjani commented:
> One issue that needs to be addressed after this PR - how do we handle LBWS conversion without duplicating tons of code? The LBWS manifest is almost exactly the same as the Backend Service manifest, but they use two different data structures in our codebase because the Port field is optional on Backend services. Alternatively, we can just not add the same level of unit tests for the LBWS conversion function since the logic should otherwise be the same.

We do plan to refactor that package to reduce the duplicated code. One day...!

rclinard-amzn committed:
Update internal/pkg/docker/dockercompose/svc_image.go

Co-authored-by: Janice Huang <60631893+huanjani@users.noreply.github.com>

huanjani commented on internal/pkg/docker/dockercompose/svc.go:
```diff
@@ -0,0 +1,104 @@
+	"time"
+)
+
+func convertBackendService(service *types.ServiceConfig, port uint16) (*manifest.BackendServiceConfig, IgnoredKeys, error) {
```
Does the `compose-go` library handle `port` separately? Where is port coming from? Is that from `expose`?

rclinard-amzn replied:
There is some complex logic that will be needed to derive the proper value for the `port` that will happen in Milestone 3 - since our model of a service having a single port doesn't exactly line up with how Compose works with its port declarations. It is being passed in here since that logic will also decide whether the service becomes a Backend Service or a Load Balanced Web Service, so it needs to be called outside of `convertBackendService` anyways.

huanjani commented on internal/pkg/docker/dockercompose/svc_image.go:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	if badArgs != nil {
+		return nil, fmt.Errorf("some entries are missing values and require user input, "+
+			"this is not supported in Copilot: %v", badArgs)
```
maybe?: 
```suggestion
		return nil, fmt.Errorf("%s '%v' %s missing %s and %s user input", english.PluralWord(badArgs, "entry", "entries"), badArgs, english.PluralWord(badArgs, "is", "are"), english.PluralWord(badArgs, "a value", "values"), english.PluralWord(badArgs, "requires", "require"))
```

rclinard-amzn replied:
Looks like a nice change! I'll keep the "not supported" part though I think.

rclinard-amzn committed:
chore: update missing values error message

Co-authored-by: Janice Huang <60631893+huanjani@users.noreply.github.com>

rclinard-amzn committed:
chore: fix compile error in previous commit and add "this is unsupported"

rclinard-amzn committed:
chore: fix tests

iamhopaul123 commented on internal/pkg/docker/dockercompose/unsupported.go:
```diff
@@ -0,0 +1,79 @@
+
+// NOTE: This file is currently unused, it will be used in the next PR. Don't mind the comments for now!
+
+type IgnoredKeys []string
```
Comment for public type pls 🙏  Also what do we use it for?

dannyrandall replied:
does it need to be exported...? (probably related to my other comment about trying to understand how this will all be used)

rclinard-amzn replied:
Comment added! And this does need to be exported since it will eventually be threaded through all the way to the CLI output.

iamhopaul123 commented on internal/pkg/docker/dockercompose/svc.go:
```diff
@@ -0,0 +1,104 @@
+func convertBackendService(service *types.ServiceConfig, port uint16) (*manifest.BackendServiceConfig, IgnoredKeys, error) {
+	image, ignored, err := convertImageConfig(service.Build, service.Labels, service.Image)
+	if err != nil {
+		return nil, ignored, fmt.Errorf("convert image config: %w", err)
```
```suggestion
		return nil, nil, fmt.Errorf("convert image config: %w", err)
```
If it fails at some point we probably want to fail entirely and in that scenario we won't use `ignored` anymore.

iamhopaul123 commented on internal/pkg/docker/dockercompose/svc_image.go:
```diff
@@ -0,0 +1,144 @@
+	}
+
+	if len(ignoredKeys) == 0 {
+		return nil, nil
```
Why do we return `nil` 🤔

rclinard-amzn replied:
It made testing more convenient (since `nil` is the empty value for maps) but I found a different way to do it that allows us to remove this.

dannyrandall commented on internal/pkg/docker/dockercompose/svc.go:
```diff
@@ -0,0 +1,104 @@
+	"time"
+)
+
+func convertBackendService(service *types.ServiceConfig, port uint16) (*manifest.BackendServiceConfig, IgnoredKeys, error) {
```
should this be an exported function? i'm trying to understand how a future command would use this package...

rclinard-amzn replied:
This will actually be called by another function that will be exported instead! This is an internal helper for those functions, since it doesn't handle actually parsing the Compose yaml.

dannyrandall commented on internal/pkg/docker/dockercompose/svc_image.go:
```diff
@@ -0,0 +1,144 @@
+		DockerLabels: labels,
+	}
+
+	// note: Compose allows both build & image to be specified, but image takes precedence.
```
nit
```suggestion
	// Compose allows both build & image to be specified, but image takes precedence.
```

dannyrandall commented on internal/pkg/docker/dockercompose/svc_image.go:
```diff
@@ -0,0 +1,144 @@
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/copilot-cli/internal/pkg/manifest"
+	"github.com/compose-spec/compose-go/types"
```
while in general i avoid this...i wonder if renaming this import would help with readability💭 something like
```suggestion
	compose "github.com/compose-spec/compose-go/types"
```
and then references would look like `compose.BuildConfig` against `manifest.Image`. idk, just a thought

rclinard-amzn replied:
Makes sense... I did this and it did improve readability, so I'll push it, thanks for the suggestion!

dannyrandall commented on internal/pkg/docker/dockercompose/svc_image.go:
```diff
@@ -0,0 +1,144 @@
+	}
+
+	if build.ExtraHosts != nil {
+		return nil, errors.New("key `build.extra_hosts` is not supported yet, this might break your app")
```
:joy:

dannyrandall commented on internal/pkg/docker/dockercompose/svc_image.go:
```diff
@@ -0,0 +1,144 @@
+		ignoredKeys = append(ignoredKeys, "build.labels")
+	}
+
+	if len(ignoredKeys) == 0 {
+		return nil, nil
+	}
```
i think we can probably just remove the `if` block...

dannyrandall commented on internal/pkg/docker/dockercompose/svc_image.go:
```diff
@@ -0,0 +1,144 @@
+	var badArgs []string
+
+	for k, v := range inArgs {
+		if v == nil {
+			badArgs = append(badArgs, k)
+		} else {
+			args[k] = *v
+		}
```
nit
```suggestion
		if v == nil {
			badArgs = append(badArgs, k)
			continue
		}
		args[k] = *v
```

dannyrandall reviewed (commented):
logic looks pretty good! just a few questions & nits!

rclinard-amzn committed:
chore: Explain why IgnoredKeys exists

rclinard-amzn committed:
chore: don't return non-nil ignored value on error

rclinard-amzn committed:
chore: don't return nil to mean empty from convertMappingWithEquals

rclinard-amzn committed:
chore: remove "note:" prefix in comments

rclinard-amzn committed:
chore: use continue instead of else

Co-authored-by: danny randall <10566468+dannyrandall@users.noreply.github.com>

rclinard-amzn committed:
chore: use "compose" package alias for code readability

Thanks Danny!

rclinard-amzn committed:
Merge remote-tracking branch 'origin/experiment/docker-compose-start' into experiment/docker-compose-start

rclinard-amzn commented:
Thanks for the great feedback, that should all be addressed now!

huanjani commented on internal/pkg/docker/dockercompose/svc_image.go:
```diff
@@ -0,0 +1,143 @@
+func nilIfEmpty(s string) *string {
+	if s == "" {
+		return nil
+	} else {
```
nit:
```suggestion
	}
```

huanjani commented on internal/pkg/docker/dockercompose/svc_image.go:
```diff
@@ -0,0 +1,143 @@
+func nilIfEmpty(s string) *string {
+	if s == "" {
+		return nil
+	} else {
+		return &s
+	}
```
nit:
```suggestion
	}
	return &s
```

rclinard-amzn committed:
Update internal/pkg/docker/dockercompose/svc_image.go

Co-authored-by: Janice Huang <60631893+huanjani@users.noreply.github.com>

rclinard-amzn committed:
chore: actually remove the "return nil, nil" part

rclinard-amzn committed:
Merge remote-tracking branch 'origin/experiment/docker-compose-start' into experiment/docker-compose-start

huanjani commented on internal/pkg/docker/dockercompose/svc_image.go:
```diff
@@ -0,0 +1,143 @@
+	}
+
+	if len(badArgs) != 0 {
+		return nil, fmt.Errorf("%s '%v' %s missing %s and %s user input, this is unsupported in Copilot",
```
What do you think about this (plus deleting "requires"/"require"? We get across the idea that this is not allowed in Copilot (even if it is in Docker Compose), but is less redundant.
```suggestion
		return nil, fmt.Errorf("%s '%v' %s missing %s; this is unsupported in Copilot",
```

rclinard-amzn replied:
Sure, that makes sense - I wanted to error on the side of not being too vague, but I think that can certainly be removed. Thanks!

huanjani reviewed (approved):
LGTM!

rclinard-amzn committed:
nit: reduce redundancy in error message

Co-authored-by: Janice Huang <60631893+huanjani@users.noreply.github.com>

rclinard-amzn committed:
chore: fix previous commit

rclinard-amzn committed:
chore: remove unnecessary nil check

rclinard-amzn committed:
chore: remove incorrect Extensions field handling

compose-go already checks for these fields and throws an error, we have no need to do so.

dannyrandall commented on internal/pkg/docker/dockercompose/svc.go:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	cmd := healthcheck.Test
+
```
```suggestion
```

dannyrandall commented on internal/pkg/docker/dockercompose/svc.go:
```diff
@@ -0,0 +1,100 @@
+// convertHealthCheckConfig trivially converts a Compose container health check into its Copilot variant.
+func convertHealthCheckConfig(healthcheck *compose.HealthCheckConfig) manifest.ContainerHealthCheck {
+	retries := 3
+
```
```suggestion
```

dannyrandall reviewed (approved):
lgtm!

rclinard-amzn committed:
chore: remove unnecessary whitespace in convertHealthCheckCOnfig

Co-authored-by: danny randall <10566468+dannyrandall@users.noreply.github.com>
